shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -0,0 +1,352 @@
+        default:
+          // Present default key if bad input type is given.
+          return EditorInfo.IME_ACTION_UNSPECIFIED;
+      }
+    }
+
+    public final boolean obscureText;
+    public final boolean autocorrect;
+    public final TextCapitalization textCapitalization;
+    public final InputType inputType;
+    public final Integer inputAction;
+    public final String actionLabel;
+
+    public Configuration(
+      boolean obscureText,
+      boolean autocorrect,
+      @NonNull TextCapitalization textCapitalization,
+      @NonNull InputType inputType,
+      @Nullable Integer inputAction,
+      @Nullable String actionLabel
+    ) {
+      this.obscureText = obscureText;
+      this.autocorrect = autocorrect;
+      this.textCapitalization = textCapitalization;
+      this.inputType = inputType;
+      this.inputAction = inputAction;
+      this.actionLabel = actionLabel;
+    }
+  }
+
+  public static class InputType {
+    public static InputType fromJson(JSONObject json) throws JSONException {
+      return new InputType(
+        TextInputType.fromValue(json.getString("name")),
+        json.optBoolean("signed", false),
+        json.optBoolean("decimal", false)
+      );
+    }
+
+    public final TextInputType type;
+    public final boolean isSigned;
+    public final boolean isDecimal;
+
+    public InputType(@NonNull TextInputType type, boolean isSigned, boolean isDecimal) {
+      this.type = type;
+      this.isSigned = isSigned;
+      this.isDecimal = isDecimal;
+    }
+  }
+
+  public enum TextInputType {
+    DATETIME("TextInputType.datetime"),
+    NUMBER("TextInputType.number"),
+    PHONE("TextInputType.phone"),
+    MULTILINE("TextInputType.multiline"),
+    EMAIL_ADDRESS("TextInputType.emailAddress"),
+    URL("TextInputType.url");
+
+    static TextInputType fromValue(String encodedName) {
+      for (TextInputType textInputType : TextInputType.values()) {
+        if (textInputType.encodedName.equals(encodedName)) {
+          return textInputType;
+        }
+      }
+      throw new RuntimeException("No such TextInputType: " + encodedName);
+    }
+
+    private final String encodedName;
+
+    TextInputType(String encodedName) {
+      this.encodedName = encodedName;
+    }
+  }
+
+  public enum TextCapitalization {
+    CHARACTERS("TextCapitalization.characters"),
+    WORDS("TextCapitalization.words"),
+    SENTENCES("TextCapitalization.sentences");
+
+    static TextCapitalization fromValue(String encodedName) {
+      for (TextCapitalization textCapitalization : TextCapitalization.values()) {
+        if (textCapitalization.encodedName.equals(encodedName)) {
+          return textCapitalization;
+        }
+      }
+      throw new RuntimeException("No such TextCapitalization: " + encodedName);
+    }
+
+    private final String encodedName;
+
+    TextCapitalization(String encodedName) {
+      this.encodedName = encodedName;
+    }
+  }
+
+  public static class TextEditState {
```

Comment:
Do we not need to model the composing region in this class?

Reply:
I don't think I know the context around that question.  Can you elaborate?